pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java

```diff
@@ -240,6 +244,13 @@ public CompletableFuture<Void> checkIfTBRecoverCompletely(boolean isTxnEnabled)
     @Override
     public CompletableFuture<Position> appendBufferToTxn(TxnID txnId, long sequenceId, ByteBuf buffer) {
         CompletableFuture<Position> completableFuture = new CompletableFuture<>();
+        Long lowWaterMark = lowWaterMarks.get(txnId.getMostSigBits());
+        if (lowWaterMark != null && lowWaterMark >= txnId.getLeastSigBits()) {
+            completableFuture.completeExceptionally(new TransactionBufferException
+                    .TransactionNotFoundException("Transaction [" + txnId + "] has been ended. "
+                    + "Please use a new transaction to send message."));
+            return completableFuture;
```

Comment:
Please check here https://github.com/apache/pulsar/blob/4f53e8eedc8aca406f60ab4c1f7bc11577a6bf20/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java#L2977-L2984

The TransactionNotFoundException will be wrapped with ManagedLedgerException. It should be wrapped by PersistenceException directly?

Reply:
I have change exception from `TransactionNotFoundException` to `NotAllowedException`.
And the `NotAllowedException` is not only used to `Exceed maximum message size`, so i change  the `recoverNotAllowedError` to make the expression of the log clearer.